src/diffusers/models/attention.py

```diff
@@ -340,12 +345,23 @@ def forward(self, hidden_states):
         key_proj = self.key(hidden_states)
         value_proj = self.value(hidden_states)
 
+        if encoder_hidden_states is not None:
+            encoder_hidden_states = encoder_hidden_states.transpose(1, 2)
+            context_key_proj = self.context_key(encoder_hidden_states)
+            context_value_proj = self.context_value(encoder_hidden_states)
+
         scale = 1 / math.sqrt(self.channels / self.num_heads)
 
         query_proj = self.reshape_heads_to_batch_dim(query_proj)
         key_proj = self.reshape_heads_to_batch_dim(key_proj)
         value_proj = self.reshape_heads_to_batch_dim(value_proj)
 
+        if encoder_hidden_states is not None:
+            context_key_proj = self.reshape_heads_to_batch_dim(context_key_proj)
+            context_value_proj = self.reshape_heads_to_batch_dim(context_value_proj)
+            key_proj = torch.concat([key_proj, context_key_proj], dim=1)
```

Comment:
That's very interesting! Seems like quite a special cross attention mechanism. Ok to do this way for now. Note that mid-term we would probably like to completely remove this class and instead make one big Attention class out of this one: https://github.com/huggingface/diffusers/blob/62b497c418d77c78215e3258a58e5da85cba1310/src/diffusers/models/attention.py#L496 

Ok for now though as we'll be able to change it back later.

Reply:
Ok cool, yes agreed merging the two blocks makes sense!